plugins/wasm-go/pkg/wrapper/redis_wrapper.go

```diff
@@ -139,39 +153,68 @@ func RedisCall(cluster Cluster, respQuery []byte, callback RedisResponseCallback
-			if status != 0 {
-				proxywasm.LogCriticalf("Error occured while calling redis, it seems cannot connect to the redis cluster. request-id: %s", requestID)
-				responseValue = resp.ErrorValue(fmt.Errorf("cannot connect to redis cluster"))
+		redisRespCallback(cluster, respQuery, callback, requestID))
+	if err != nil {
+		proxywasm.LogCriticalf("redis call failed, request-id: %s, error: %v", requestID, err)
+	} else {
+		proxywasm.LogDebugf("redis call start, request-id: %s, respQuery: %s", requestID, base64.StdEncoding.EncodeToString(respQuery))
+	}
+	return err
+}
+
+func redisRespCallback(cluster Cluster, respQuery []byte, callback RedisResponseCallback, requestID string) func(status int, responseSize int) {
+	return func(status int, responseSize int) {
+		response, err := proxywasm.GetRedisCallResponse(0, responseSize)
+		var responseValue resp.Value
+		if status != 0 {
+			proxywasm.LogCriticalf("Error occured while calling redis, it seems cannot connect to the redis cluster. request-id: %s", requestID)
+			responseValue = resp.ErrorValue(fmt.Errorf("cannot connect to redis cluster"))
+		} else {
+			if err != nil {
+				proxywasm.LogCriticalf("failed to get redis response body, request-id: %s, error: %v", requestID, err)
+				responseValue = resp.ErrorValue(fmt.Errorf("cannot get redis response"))
 			} else {
-				if err != nil {
-					proxywasm.LogCriticalf("failed to get redis response body, request-id: %s, error: %v", requestID, err)
-					responseValue = resp.ErrorValue(fmt.Errorf("cannot get redis response"))
+				rd := resp.NewReader(bytes.NewReader(response))
+				value, _, err := rd.ReadValue()
+				if err != nil && err != io.EOF {
+					proxywasm.LogCriticalf("failed to read redis response body, request-id: %s, error: %v", requestID, err)
+					responseValue = resp.ErrorValue(fmt.Errorf("cannot read redis response"))
 				} else {
-					rd := resp.NewReader(bytes.NewReader(response))
-					value, _, err := rd.ReadValue()
-					if err != nil && err != io.EOF {
-						proxywasm.LogCriticalf("failed to read redis response body, request-id: %s, error: %v", requestID, err)
-						responseValue = resp.ErrorValue(fmt.Errorf("cannot read redis response"))
-					} else {
-						responseValue = value
-						proxywasm.LogDebugf("redis call end, request-id: %s, respQuery: %s, respValue: %s",
-							requestID, base64.StdEncoding.EncodeToString([]byte(respQuery)), base64.StdEncoding.EncodeToString(response))
-					}
+					responseValue = value
+					proxywasm.LogDebugf("redis call end, request-id: %s, respQuery: %s, respValue: %s",
+						requestID, base64.StdEncoding.EncodeToString([]byte(respQuery)), base64.StdEncoding.EncodeToString(response))
 				}
 			}
-			if callback != nil {
-				callback(responseValue)
+		}
+		key := "redisInitialized" + requestID
+		data, cas, err := proxywasm.GetSharedData(key)
+		if strings.HasPrefix(responseValue.String(), "NOAUTH") && (errors.Is(err, types.ErrorStatusNotFound) || data == nil) {
```

Comment:
The logic here seems incorrect as well; I think you want to reinitialize and retry only when data is nil and there is a NOAUTH error